api/src/main/java/uk/gov/dvsa/mot/motr/service/MotrVehicleHistoryProvider.java

```diff
@@ -0,0 +1,140 @@
+package uk.gov.dvsa.mot.motr.service;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import uk.gov.dvsa.mot.motr.model.HgvPsvVehicleWithLatestTest;
+import uk.gov.dvsa.mot.motr.model.VehicleWithLatestTest;
+import uk.gov.dvsa.mot.trade.api.InvalidResourceException;
+import uk.gov.dvsa.mot.trade.api.TradeException;
+import uk.gov.dvsa.mot.vehicle.hgv.HgvVehicleProvider;
+import uk.gov.dvsa.mot.vehicle.hgv.model.Vehicle;
+import uk.gov.dvsa.mot.vehicle.hgv.validation.TrailerIdFormat;
+
+import java.util.Objects;
+import java.util.Optional;
+
+import javax.inject.Inject;
+import javax.validation.constraints.NotNull;
+
+public class MotrVehicleHistoryProvider {
+
+    private static final Logger logger = LogManager.getLogger(MotrVehicleHistoryProvider.class);
+
+    private MotrReadService motrReadService;
+    private HgvVehicleProvider hgvVehicleProvider;
+
+    @Inject
+    public MotrVehicleHistoryProvider(MotrReadService motrReadService, HgvVehicleProvider hgvVehicleProvider) {
+        this.motrReadService = motrReadService;
+        this.hgvVehicleProvider = hgvVehicleProvider;
+    }
+
+    @NotNull public VehicleWithLatestTest searchVehicleByRegistration(@NotNull String registration) throws Exception {
+        Optional<VehicleWithLatestTest> optionalMotVehicle = motrReadService.getLatestMotTestByRegistration(registration);
+
+        if (optionalMotVehicle.isPresent()) {
+            return getVehicleDetailsFromMostRecentSource(optionalMotVehicle.get());
+        }
+
+        Optional<VehicleWithLatestTest> optionalDvlaVehicle = motrReadService.getLatestMotTestForDvlaVehicleByRegistration(registration);
+
+        if (shouldGetHgvPsvHistory(optionalDvlaVehicle.isPresent(), registration)) {
+
+            Optional<Vehicle> optionalHgvPsvVehicle = Optional.ofNullable(hgvVehicleProvider.getVehicle(registration));
+
+            if (optionalHgvPsvVehicle.isPresent()) {
+                return new HgvPsvVehicleWithLatestTest(optionalHgvPsvVehicle.get(),
+                        optionalDvlaVehicle.map(VehicleWithLatestTest::getDvlaVehicleId).orElse(null));
+            } else {
+                logger.debug("No HGV/PSV vehicle retrieved");
```

Comment:
Might be nice to add some more steps in logging chain (after trying to find mot vehicle, after trying to find dvla vehicle etc)

Reply:
added